### ISSUE TITLE
reportDir is not working

### ISSUE THREAD

**Issue by n0m0r3pa1n** [koral--/gradle-pitest-plugin#63]
I tried to override the reportDir both while configuring the plugin and again through command line but it seems that it is always generated in app/build/reports/pitest/

This is what I have tried:
./gradlew pitestDebug  -Doverride.pitest.reportDir=./app/target/pit-reports/
./gradlew pitestDebug  -Doverride.pitest.report.dir=./app/target/pit-reports/

```
configure<pl.droidsonroids.gradle.pitest.PitestPluginExtension> {
    val reportFileDir = file("${buildDir.parentFile}/target/pit-reports")
    if (!reportFileDir.exists()) {
        reportFileDir.mkdir()
    }
    this.setReportDir(reportFileDir)
    targetClasses.set(listOf("com.****.*****.*"))
    targetTests.set(listOf("com.*****.*****.*"))
    excludedClasses.set(
        listOf(
            "com.*****.****.databinding.*",
            "*.views.*",
            "*.activities.*",
            "*.fragments.*",
            "*.events.*",
            "*.di.*",
            "*.model.*",
            "*.models.*"
        )
    )
    threads.set(4)
    verbose.set(true)
    timestampedReports.set(false)
    outputFormats.set(listOf("XML"))
}
```

**Comment by szpak** [comment]
Have you applied the [gradle-override-plugin](https://github.com/nebula-plugins/gradle-override-plugin) plugin (as mentioned in the readme)?

Please call Gradle with `--info` and paste the command line which is used to execute pit process itself (you can redact the output to remove classpath, to do not clutter the output).

**Comment by n0m0r3pa1n** [comment]
So I did apply the plugin, but the issue continues.
```
./gradlew pitestDebug -Doverride.pitest.reportDir=target/pit-reports

> Configure project :app
Dex options: javaMaxHeapSize = 4096m

> Configure project :
Using config.yml
Applying build configuration from file:config.yml
Using debug KeyStore
Test results are empty

FAILURE: Build failed with an exception.

* Where:
Build file '/home/nmp/Workspace/numbrs/build.gradle.kts' line: 202

* What went wrong:
A problem occurred configuring project ':app'.
> Cannot set the value of a property of type org.gradle.api.file.Directory using an instance of type java.lang.String.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 916
```

**Comment by szpak** [comment]
I will not work. I forgot I changed `String` to `Directory` (even `Property<Directory>`?) in the plugin configuration - to be more compatible with Gradle 7/8 and the override plugin doesn't support it. AFAIR `@Option` also doesn't work with `Directory`. I will update the README to reflect that.

As a workaround, you can pass your own `-Pxxx.pitest.reportDir=target/pit-reports`, read it with `project.findProperty()` in `build.gradle` and set `reportDir` property in code, if `-P` is provided.

**Comment by n0m0r3pa1n** [comment]
Okay but are you sure that it will work when I do it this way? I am asking because when I configure the plugin inside my app build.gradle.kts it does basically nothing:
```
configure<pl.droidsonroids.gradle.pitest.PitestPluginExtension> {
    val reportFileDir = file("${buildDir.parentFile}/target/pit-reports")
    if (!reportFileDir.exists()) {
        reportFileDir.mkdir()
    }
    this.setReportDir(reportFileDir)
    targetClasses.set(listOf("com.****.*****.*"))
    targetTests.set(listOf("com.*****.*****.*"))
    excludedClasses.set(
        listOf(
            "com.*****.****.databinding.*",
            "*.views.*",
            "*.activities.*",
            "*.fragments.*",
            "*.events.*",
            "*.di.*",
            "*.model.*",
            "*.models.*"
        )
    )
    threads.set(4)
    verbose.set(true)
    timestampedReports.set(false)
    outputFormats.set(listOf("XML"))
}
```
In the Gradle output I see that the targetClasses, excludedClasses, threads etc are picked up (they are printed in the ReportOptions), the only thing that is not picked up is the path for the reportDir. I am wondering whether this thing will be picked up even with a custom parameter.

**Comment by n0m0r3pa1n** [comment]
Isn't the directory path set static here:
https://github.com/koral--/gradle-pitest-plugin/blob/master/src/main/groovy/pl/droidsonroids/gradle/pitest/PitestPlugin.groovy#L75

I think it is not taken from the configuration?

**Comment by szpak** [comment]
Hmm, it seems that @koral-- who maintained the Android fork of gradle-pitest-plugin (this project) made some changes in comparison to the original project. My proposal should work in the original project, but with the line you pasted it is more problematic. `reportDir` is [taken](https://github.com/koral--/gradle-pitest-plugin/blob/53cff8bd33c432105fb752f8f64033345a9f3af8/src/main/groovy/pl/droidsonroids/gradle/pitest/PitestPlugin.groovy#L170) from the configuration, but that `afterEvaluate` seems to be executed later on. @koral-- what is the purpose of that line? Is it a [workaround](https://github.com/szpak/gradle-pitest-plugin/blob/142ede77a369c2a321fba4656f7a84f405e2d3be/src/main/groovy/info/solidsoft/gradle/pitest/PitestPlugin.groovy#L123-L126) on Gradle <6 compatibility?

@n0m0r3pa1n Sorry for confusion. I haven't checked that related mechanism has been changed in the fork.

**Comment by koral--** [comment]
@szpak it is taken directly from upstream
https://github.com/szpak/gradle-pitest-plugin/commit/3f431599111ea5d2a7f19503b428406213b13770

**Comment by szpak** [comment]
@koral-- Yes, and that's correct :). I wanted to ask about that [line](
https://github.com/koral--/gradle-pitest-plugin/commit/53cff8bd33c432105fb752f8f64033345a9f3af8#diff-b6ed157923d9b558458dddb42ef25ea0R75).

In the upstream version, I [set](https://github.com/szpak/gradle-pitest-plugin/blob/142ede77a369c2a321fba4656f7a84f405e2d3be/src/main/groovy/info/solidsoft/gradle/pitest/PitestPlugin.groovy#L106) the default value when plugin is applied, but I allow user to set anything he/she wants in the configuration. I you really want to change it back to the default value if user changed it to null (or rather an empty provider), it should be done conditionally. The `if` has been (probably accidentally) removed in the first link I pasted. However, in there are no other specific corner cases, it would consider just the removal of that line, as it should generate nice error message about not set value on of the mandatory field `reportDir` at the task level anyway.

**Comment by szpak** [comment]
Hmm, for unknown reason, I'm not able to get that nice [code preview](https://help.github.com/en/github/managing-your-work-on-github/creating-a-permanent-link-to-a-code-snippet) in the issues...

https://github.com/koral--/gradle-pitest-plugin/pull/59/files#diff-b6ed157923d9b558458dddb42ef25ea0R75-L79

https://github.com/koral--/gradle-pitest-plugin/blob/5a40193b312f1adf87e1ffae0f06cd8b923e561a/src/main/groovy/pl/droidsonroids/gradle/pitest/PitestPlugin.groovy#L75

**Comment by koral--** [comment]
Sorry for my slow responsiveness.
Regarding `afterEvaluate`, it was added because in Android fork we have separate pitest task for each build variant (by default debug and release but each build type and flavor dimension combination adds one) and variant list is created lazily.

If I understand this issue correctly it is just caused by the fact that line setting report path wasn't updated with upstream for a long time. It should be resolved after merging #65.
